### PR TITLE
Updates to allow for Namespacing ACL resources in Consul Enterp…

### PR DESCRIPTION
**Pull request by mkeeler**
This PR is for the necessary modifications to Consul OSS to enable Consul Enterprise Namespacing of ACL resources.

## Main Changes

* method signature updates everywhere to account for passing around enterprise meta.
* populate the EnterpriseAuthorizerContext for all ACL related authorizations.
* ACL resource listings now operate like the catalog or kv listings in that the returned entries are filtered down to what the token is allowed to see. With Namespaces its no longer all or nothing.
* Modified the acl.Policy parsing to abstract away basic decoding so that enterprise can do it slightly differently. Also updated method signatures so that when parsing a policy it can take extra ent metadata to use during rules validation and policy creation.

## Secondary Changes

* Moved protobuf encoding functions out of the agentpb package to eliminate circular dependencies.
* AuthMethod validator cache is now an interface as these will be cached per-namespace for Consul Enterprise.
* Added checks for policy/role link existence at the RPC API so we don’t push the request through raft to have it fail internally.
* Forward ACL token delete request to the primary datacenter when the secondary DC doesn’t have the token.
* Added a bunch of ACL test helpers for inserting ACL resource test data.

### CODE REVIEW

File: agent/consul/acl_endpoint.go [hunk 1]
```diff
@@ -715,10 +752,18 @@ func (a *ACL) TokenDelete(args *structs.ACLTokenDeleteRequest, reply *string) er
 
 		// No need to check expiration time because it's being deleted.
 
+		// token found in secondary DC buts its not local so it must be deleted in the primary
```
Comment (rboyer) [inline agent/consul/acl_endpoint.go]:
```suggestion
		// token found in secondary DC but it's not local so it must be deleted in the primary
```

File: agent/consul/acl_endpoint.go [hunk 2]
```diff
@@ -715,10 +752,18 @@ func (a *ACL) TokenDelete(args *structs.ACLTokenDeleteRequest, reply *string) er
...
 		if !a.srv.InACLDatacenter() && !token.Local {
 			args.Datacenter = a.srv.config.ACLDatacenter
 			return a.srv.forwardDC("ACL.TokenDelete", a.srv.config.ACLDatacenter, args, reply)
 		}
+	} else if !a.srv.InACLDatacenter() {
+		// token not found in secondary DC - attempt to delete within the primary
+		args.Datacenter = a.srv.config.ACLDatacenter
+		return a.srv.forwardDC("ACL.TokenDelete", a.srv.config.ACLDatacenter, args, reply)
+	} else {
+		// in Primary Datacenter but the token does not exists - return early as there is nothing to do.
```
Comment (rboyer) [inline agent/consul/acl_endpoint.go]:
```suggestion
		// in Primary Datacenter but the token does not exist - return early as there is nothing to do.
```

File: agent/consul/server.go [hunk 3]
```diff
@@ -114,8 +115,7 @@ type Server struct {
 	// acls is used to resolve tokens to effective policies
 	acls *ACLResolver
 
-	aclAuthMethodValidators    map[string]*authMethodValidatorEntry
-	aclAuthMethodValidatorLock sync.RWMutex
+	aclAuthMethodValidators authmethod.AuthMethodCache
```
Comment (rboyer) [inline agent/consul/server.go]: trivial: maybe rename `authmethod.AuthMethodCache` to `authmethod.Cache` to remove the identifier stutter

Comment (mkeeler) [reply agent/consul/server.go]: will do

File: agent/consul/state/acl.go [hunk 4]
```diff
@@ -2107,48 +1823,50 @@ func (s *Store) ACLAuthMethodList(ws memdb.WatchSet) (uint64, structs.ACLAuthMet
...
+		// Therefore this is fine but if we ever change that precondition then this will be wrong (unless
+		// we ensure all deletions in a batch should have the same enterprise meta)
+		s.aclAuthMethodDeleteTxn(tx, idx, name, entMeta)
 	}
 
-	if err := indexUpdateMaxTxn(tx, idx, "acl-auth-methods"); err != nil {
-		return fmt.Errorf("failed updating index: %v", err)
-	}
+	//
+	// TODO figure out how to modify the batch deletion struct to blow away the auth methods.
```
Comment (rboyer) [inline agent/consul/state/acl.go]: Did you want to use an annotated TODO here?

File: agent/consul/state/acl_oss.go [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/agent/structs"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+///////////////////////////////////////////////////////////////////////////////
+/////                          ACL Table Schemas                          /////
```
Comment (rboyer) [inline agent/consul/state/acl_oss.go]: These should be identical to the schemas removed from the other file right?

File: agent/consul/state/acl_oss.go [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+
+func (s *Store) aclTokenMaxIndex(tx *memdb.Txn, _ *structs.ACLToken, entMeta *structs.EnterpriseMeta) uint64 {
+	return maxIndexTxn(tx, "acl-tokens")
+}
+
+func (s *Store) aclTokenUpsertValidateEnterprise(tx *memdb.Txn, token *structs.ACLToken, existing *structs.ACLToken) error {
+	return nil
+}
+
+func (s *Store) ACLTokenUpsertValidateEnterprise(token *structs.ACLToken, existing *structs.ACLToken) error {
```
Comment (rboyer) [inline agent/consul/state/acl_oss.go]: Why do both `aclTokenUpsertValidateEnterprise` and `ACLTokenUpsertValidateEnterprise` exist?

File: agent/http_register_oss.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	registerEndpoint("/v1/agent/host", []string{"GET"}, (*HTTPServer).AgentHost)
+	registerEndpoint("/v1/agent/maintenance", []string{"PUT"}, (*HTTPServer).AgentNodeMaintenance)
+	registerEndpoint("/v1/agent/reload", []string{"PUT"}, (*HTTPServer).AgentReload)
+	registerEndpoint("/v1/agent/monitor", []string{"GET"}, (*HTTPServer).AgentMonitor)
+	registerEndpoint("/v1/agent/metrics", []string{"GET"}, (*HTTPServer).AgentMetrics)
+	registerEndpoint("/v1/agent/services", []string{"GET"}, (*HTTPServer).AgentServices)
+	registerEndpoint("/v1/agent/service/", []string{"GET"}, (*HTTPServer).AgentService)
+	registerEndpoint("/v1/agent/checks", []string{"GET"}, (*HTTPServer).AgentChecks)
+	registerEndpoint("/v1/agent/members", []string{"GET"}, (*HTTPServer).AgentMembers)
+	registerEndpoint("/v1/agent/join/", []string{"PUT"}, (*HTTPServer).AgentJoin)
```
Comment (rboyer) [inline agent/http_register_oss.go]: I guess this is more of a general comment about tracking things that have alter egos. Why is this file suffixed with `_oss.go` if it is not conditionally compiled over in enterprise? My initial read when I saw this was "oh no, all of these are duplicated elsewhere" but I guess they aren't?

File: agent/structs/acl.go [hunk 8]
```diff
@@ -1169,6 +1329,8 @@ type ACLPolicyListResponse struct {
 type ACLPolicyBatchGetRequest struct {
 	PolicyIDs  []string // List of policy ids to fetch
 	Datacenter string   // The datacenter to perform the request within
+	// TODO (namespaces) should we include ent meta? Policy IDs will have
+	// to be unique acorss namespaces so in theory it should be unnecessary.
```
Comment (rboyer) [inline agent/structs/acl.go]:
```suggestion
	// to be unique across namespaces so in theory it should be unnecessary.
```

File: agent/structs/acl.go [hunk 9]
```diff
@@ -1201,6 +1363,8 @@ type ACLPolicyBatchSetRequest struct {
 // This is particularly useful during replication
 type ACLPolicyBatchDeleteRequest struct {
 	PolicyIDs []string
+	// TODO (namespaces) should we include ent meta? Policy IDs will have
+	// to be unique acorss namespaces so in theory it should be unnecessary.
```
Comment (rboyer) [inline agent/structs/acl.go]:
```suggestion
	// to be unique across namespaces so in theory it should be unnecessary.
```

File: agent/structs/acl.go [hunk 10]
```diff
@@ -1267,6 +1434,8 @@ type ACLRoleListResponse struct {
 type ACLRoleBatchGetRequest struct {
 	RoleIDs    []string // List of role ids to fetch
 	Datacenter string   // The datacenter to perform the request within
+	// TODO (namespaces) should we include ent meta? Role IDs will have
+	// to be unique acorss namespaces so in theory it should be unnecessary.
```
Comment (rboyer) [inline agent/structs/acl.go]:
```suggestion
	// to be unique across namespaces so in theory it should be unnecessary.
```

File: agent/structs/acl.go [hunk 11]
```diff
@@ -1300,6 +1469,8 @@ type ACLRoleBatchSetRequest struct {
 // This is particularly useful during replication
 type ACLRoleBatchDeleteRequest struct {
 	RoleIDs []string
+	// TODO (namespaces) should we include ent meta? Role IDs will have
+	// to be unique acorss namespaces so in theory it should be unnecessary.
```
Comment (rboyer) [inline agent/structs/acl.go]:
```suggestion
	// to be unique across namespaces so in theory it should be unnecessary.
```

File: agent/structs/acl.go [hunk 12]
```diff
@@ -1367,6 +1541,8 @@ type ACLBindingRuleBatchSetRequest struct {
 // multiple rule deletions
 type ACLBindingRuleBatchDeleteRequest struct {
 	BindingRuleIDs []string
+	// TODO (namespaces) should we include ent meta? Binding rule IDs will have
+	// to be unique acorss namespaces so in theory it should be unnecessary.
```
Comment (rboyer) [inline agent/structs/acl.go]:
```suggestion
	// to be unique across namespaces so in theory it should be unnecessary.
```

File: agent/structs/acl_oss.go [hunk 13]
```diff
@@ -2,6 +2,36 @@
...
+	"github.com/hashicorp/consul/acl"
+)
+
 const (
 	EnterpriseACLPolicyGlobalManagement = ""
+
+	// aclPolicyTemplateServiceIdentity is the template used for synthesizing
+	// policies for service identities.
+	aclPolicyTemplateServiceIdentity = `
+service "%[1]s" {
```
Comment (rboyer) [inline agent/structs/acl_oss.go]: I have never seen this sprintf syntax before.

File: agent/consul/state/acl_oss.go (hunk 5 above)
Comment (mkeeler) [reply agent/consul/state/acl_oss.go]: yes

File: agent/consul/state/acl_oss.go (hunk 6 above)
Comment (mkeeler) [reply agent/consul/state/acl_oss.go]: In enterprise these will validate that the namespace of the token (or other acl resource is valid and unchanging). The unexported one is used within the state package and the exported ones are used by the RPC functions.

File: agent/http_register_oss.go (hunk 7 above)
Comment (mkeeler) [reply agent/http_register_oss.go]: You are right, these are common and shouldn't have the suffix.

File: agent/consul/state/acl.go (hunk 4 above)
Comment (mkeeler) [reply agent/consul/state/acl.go]: TODO is outdated here.

File: agent/structs/acl_oss.go (hunk 13 above)
Comment (mkeeler) [reply agent/structs/acl_oss.go]: It lets you use 1 parameter in multiple places without passing that parameter multiple times.

**Pull request merged**